### DISCUSSION TITLE
Preview feature feedback: `strictUndefinedChecks`

### DISCUSSION THREAD

**Discussion by jharrell** [prisma/prisma#25271] [Feedback]
`strictUndefinedChecks` is a change to Prisma Client which disallows `undefined` values in Prisma Client queries. This feature is a direct result of [this issue](https://github.com/prisma/prisma/issues/20169).

Documentation: https://www.prisma.io/docs/orm/prisma-client/special-fields-and-types/null-and-undefined#strict-undefined-checks-preview-feature
Please share your feedback about `strictUndefinedChecks` (released in [v5.20.0 of Prisma ORM](https://github.com/prisma/prisma/releases/tag/5.20.0)) in this discussion.

If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repository.
If the feature is working well for you, please share so in a comment below or leave a 👍 on this discussion.
Need help or have questions? Join the [Prisma Discord](https://pris.ly/discord) community and ask away! We're here to support you.

**Comment by ArnaudBarre** [comment]
I find this feature useful for cases like where clauses, but I think it's way too strong for update pattern where it's common for an API to expose an endpoint schema like `{ id: string; firstName?: string | null; lastName?: string | null, ...  }` where null reset the field, undefined doesn't touch it, which allow the client to only update parts of a model. This patterns is easy to implement because I can dump my 10-15 fields input validated by zod into the data of a Prisma update.
Do you plan to have it disabled for updates?
Thanks!

**Reply by jharrell** [reply]
Thanks for your feedback! If I understand your example schema, something like `{ id: 'my-id', firstName: 'Jon' }` is valid while `{ id: 'my-id', firstName: 'Jon', lastName: undefined }` is not. 

If that's the case, then this change will not impact your case. We're leveraging the fact that "a property existing and its value is undefined" differs from "a property does not exist". We are targeting the former, not the latter.

If I have misunderstood, I'm happy to chat further as well 🙂

**Reply by ArnaudBarre** [reply]
Hum sadly zod does not make it possible to differentiate `{ id: 'my-id', firstName: 'Jon' }` and `{ id: 'my-id', firstName: 'Jon', lastName: undefined }` (this could be done via refine though).

But that's true that my frontend will most of the time send only updated fields. I need to try it, this could also works but I need to see if I can enabled `exactOptionalPropertyTypes` on my project

**Reply by EvHaus** [reply]
I also wish I could ignore this feature for `data: ...` values and only have it apply to `where: ...` clauses.

**Reply by JClackett** [reply]
God, this is gna break so much of my app - I have a few generic update functions in my app (validated by zod) and because these update functions are used in a load of places comprising of different inputs in each, I mark each field as optional, so the input type could be string or undefined, (nullable in some cases too). And so if I enable `exactOptionalPropertyTypes`, it's gna error everywhere that does this pattern. Currently I can just pass the validated output straight into prisma as I know its correct, do I now have to manually assign Prisma.Skip everywhere?

**Reply by BracketJohn** [reply]
Plus one from us on the proposal by @EvHaus:
> I also wish I could ignore this feature for data: ... values and only have it apply to where: ... clauses.

We have a similar use-case to @jharrell where our `update`-pattern allows the client to either pass a value or pass `undefined`. Passing `undefined` leads to no change. As our code base is quite big, it'll be hard to impossible for us to adopt `exactOptionalPropertyTypes` if it is enforced for _all_ Prisma fields. Having it checking just the `where: `-clause would be feasible & helpful for us to adopt (and AFAIK also resolve the problem you introduced this feature for, as per docs: https://www.prisma.io/docs/orm/prisma-client/special-fields-and-types/null-and-undefined#using-strict-undefined-checks)

**Reply by Saeid-Za** [reply]
Regarding this issue, one solution would be to first clean your input, then pass it to the prisma functions.

There are [libraries](https://github.com/Saeid-Za/cleana) for that specific use case.

Disclaimer: Im the author of the package, there are other alternatives too.

**Comment by halljus** [comment]
I wish there was an alternative to `exactOptionalPropertyTypes` for detecting these cases at build-time. We tried enabling it on a medium-sized project and there were a ton of violations unrelated to Prisma (all pertaining to other third-party libraries/integrations like Playwright, React Bootstrap, AG Grid, etc.). This is going to make migration somewhat of a pain, unfortunately.

Also, the example for skipping says `email: optionalEmail ?? Prisma.skip` but that's not quite enough, is it, as `null` still means something? We're going to have to use `email: typeof optionalEmail === 'undefined' ? Prisma.skip : optionalEmail` everywhere, right? (If not, how do we set something to `null`?)

**Reply by jharrell** [reply]
Ah, yeah that's a very good point. Thank you for the feedback! 

As of right now, yes, I think that `typeof optionalEmail` is the correct path forward. I'll bring this back to our eng team to see what we can do. 

Agreed on `exactOptionalPropertyTypes`. Unfortunately our team didn't see a reasonable path forward to handle this at build time outside of it.

**Reply by ebg1223** [reply]
I appreciate the concept and intent of this change, but enabling exactOptionalPropertyTypes is also causing very large issues in multiple enterprise projects we are maintaining. It seems lots of libraries depend on this being disabled.

Unless typescript wizards come up with another option, what about a typescript-eslint rule? This might be a compromise for existing projects, especially if combined with the ability to disable the actual runtime error.

**Reply by paulwer** [reply]
what about an additional property on the level of where, to define the required parameters, this could enable users to opt in into this feature and exspecialy when using deleteMany or similar this would add an security like layer.

**Comment by maiconcarraro** [comment]
I'm going to repost this comment from the original issue: https://github.com/prisma/prisma/issues/20169#issuecomment-2372502993 

@drewcorlin1

> The release notes say that this will be the default in the next major version. While not a blocker, that will require many changes with how my team currently uses prisma.

> It's quite common for us to use prisma to implement REST GET endpoints to list entities that can have many optional query params. For that sort of use case it'll require us to make many changes to fallback to Prisma.skip.

> Is it possible to make this behavior configurable in the future?

I have the exact same situation, we are 100% comfortable using undefined to manage the operations, it's using zod to validate inputs and some of the attributes are going to be undefined... gave a try setting the `exactOptionalPropertyTypes`, it's almost impossible to cover everything in a large project.

For new major release I'm ok to be a default a option, but it **must** be configurable to allow a painless-migration.

**Reply by jharrell** [reply]
Thanks for the feedback and for bringing the comment over from the original issue 🙏 

We understand that this ranges from a "medium" to "extra extra large" change in folks' projects. We want to make that transition as smooth as possible. If you could provide examples of where the current process is painful, that would help us develop something better moving forward.

Full disclosure: we don't exactly know how we're going to do that yet. Right now we're thinking through changes to the API, code mods, and a potential "reverse" Preview feature that could be enabled in Prisma 6. We're thinking through all of the options and will be sure to update folks when we have a better idea of what the eng team wants to do.

I know that's not a super satisfactory answer right now, but it's the one we have at the moment. Thank you again for the feedback and we're looking forward to more 🙂

**Reply by drewcorlin1** [reply]
Here's a (simplified) example of the kind of thing I do commonly that is causing issues with this change enabled
```typescript
const where =
  name || ids?.length || someOtherEntityId
    ? {
        name: { contains: name, mode: 'insensitive' as const },
        id: { in: ids },
        someOtherEntityId,
      }
    : undefined;
// this.DB() returns a prisma client
const [total, items] = await this.DB().$transaction([
  this.DB().myModel.count({ where }),
  this.DB().myModel.findMany({
    where,
    skip: offset,
    take: limit,
    orderBy: { name: 'asc' },
  }),
]);
```

This is to implement an endpoint like `GET /myModel?name=abc&id=1&id=2&someOtherEntityId=3` where those query params are optional. In reality I have more query params than this but you get the idea.

I was also caught off guard that the types let undefined through but then it fails at runtime (which I realize is because of my tsconfig settings, but noting it since I assume I'm not the only one). That, in my opinion, is a bad experience and goes against most of my experience with prisma. For 99% of my other use of prisma I know if my query typechecks then it is probably right, which is one of the reasons I'm such a big fan of prisma. Losing that experience here would be significant IMO.

**Comment by danbord** [comment]
This is great.  I was previously doing that manually by extending $allOperations  and checking if any property in the where statement was undefined as a safety net.

Only thing, would it be possible to export the Skip type so I can use it in a variable in my application.  Would use it this way:

`const mySiteFilter : string | Skip = Primsa.Skip`

for now I'm doing this instead:

`siteIdFilter: string | typeof Prisma.skip = Primsa.Skip`

which is a bit annoying


Thanks

**Comment by DoubleJ-G** [comment]
I think this is a step in the right direction, skipping the condition entirely was really unintuitive behaviour. But there are still cases where you WANT to pass in undefined and receive no results back. A use case with `findFirst` or `findUnique` with intentionally undefined values from should return no results.

I have a case where the input can be `undefined` or an ObjectId and I want to potentially search that document

```ts
const vehicle = await this.prismaService.vehicle.findFirst({
      where: { id: vehicleId },
 });
 ```
If there is no `id` passed I'd expect to get no vehicle back. Neither the preview feature or previous behaviour allows this.

**Reply by danbord** [reply]
This is where you should use `null` instead of `undefined`.  `null` is still valid and filters where the value is `NULL` in the database.  `undefined` is not permitted.  If you want undefined to be considered as null in your example you could do:

```
const vehicle = await this.prismaService.vehicle.findFirst({
      where: { id: vehicleId || null },
 });
```

**Reply by DoubleJ-G** [reply]
This doesn't work when it is an id and the type can't be null

**Comment by jekh** [comment]
Is it possible to disable generation of the Skip type entirely? It makes it difficult to use the generated types as 'normal' objects.
```ts
const input: Prisma.ItemUncheckedCreateInput = {}
// ...
input.line = input.line?.replaceAll("-", "") || null
// ^^^ Property 'replaceAll' does not exist on type 'string | Skip'.
```

**Reply by dodas** [reply]
Using `satisfies` TS keyword instead of type annotation should help here:
```ts
const input = {} satisfies Prisma.ItemUncheckedCreateInput;
```

**Comment by paulwer** [comment]
Prisma.skip should work for take,skip,orderBy and OR/AND clauses as well❤️

**Comment by SerenModz21** [comment]
I've had one problem with `Prisma.skip`: I can't provide it to the cursor when requesting initial data (where the cursor is currently unknown).

For example: 
```ts
const data = await prisma.model.findMany({
    cursor: hasCursor ? { /* provided cursor info */ } : Prisma.skip,
});
```

**Reply by paulwer** [reply]
we are using js deconstructor as a workaround for now:

...hasSkip ? { skip:...} : {},

**Comment by kolliss** [comment]
"strictUndefinedChecks is a change to Prisma Client which disallows undefined values in Prisma Client queries" – that's very nice, but I'm pretty sure that this check should not apply to the content of JSON fields.

**Reply by dodas** [reply]
+1
This is extra annoying since typescript doesn't complain about this:
```ts
await prisma.person.update({
  where: {
    id,
  },
  data: {
    // `emails` and `addresses` are both JSON fields
    emails: undefined, // Type 'undefined' is not assignable to type 'Skip | { ifUndefined: {}; } | InputJsonValue | JsonNull | { readonly [x: string]: Exact<InputJsonValue | null | undefined, InputJsonValue | null | undefined>; } | { ...; } | { ...; } | {}
    addresses: [undefined], // No TS error here, but causes a prisma runtime error
  },
});
```

**Comment by basememara** [comment]
This is a great addition for safety. Not only for deletes, but even when fetching records for the user, it may return all records for the user unintendedly.

I just want to make sure though, this will only be enforced for the `where` clause? Because sometimes this is useful for creating and updating records to pass undefined to ignore the column.

**Comment by rivatove** [comment]
The change is great. If there was a way to skip it (pun intended) for update/create queries it would be perfect.

**Comment by rhyek** [comment]
I agree with another commenter that we should be able to enable this for where clauses only if so desired. My main concern when looking for something like this is to prevent accidental data loss. This is something that could affect all rows in a table and it would be great for Prisma to have protections from that.

**Reply by rhyek** [reply]
I wrote a middleware that I think will be sufficient:

```ts
  client.$use(async (params, next) => {
    const checkForUndefined = (obj: any) => {
      for (const key in obj) {
        if (obj[key] === undefined) {
          throw new Error(
            `Explicitly \`undefined\` values are not allowed in where clause (key: ${key})`
          );
        }
        if (obj[key]) {
          if (obj[key].constructor === Object) {
            checkForUndefined(obj[key]);
          } else if (Array.isArray(obj[key])) {
            for (const item of obj[key]) {
              checkForUndefined(item);
            }
          }
        }
      }
    };
    const checkWhere = (obj: any) => {
      if (obj?.where && obj.where.constructor === Object) {
        if (Object.keys(obj.where).length === 0) {
          throw new Error('Empty where clause is not allowed');
        }
        checkForUndefined(obj.where);
      }
    };
    checkWhere(params.args);
    const checkIncludes = (obj: any) => {
      if (obj?.include && obj.include.constructor === Object) {
        for (const relation of Object.values(obj.include)) {
          checkWhere(relation);
          checkIncludes(relation);
        }
      }
    };
    checkIncludes(params.args);

    return next(params);
  });
```

**Comment by wmadden** [comment]
A quick update: we're currently evaluating whether / how to proceed with this preview feature. We’ll publish an updated plan later this year. Read more here: https://github.com/prisma/prisma/discussions/26136

**Comment by biw** [comment]
I wanted to add some thoughts here after migrating our codebase to use `strictUndefinedChecks`

1. This is a fantastic feature. It has helped us catch multiple bugs that I'm not sure how we would have found without this feature
2. As others have mentioned, the support around the typescript ecosystem is not fantastic for `exactOptionalPropertyTypes`, with many 3rd party packages not making clear the difference between `?: value` and `?: value | undefined`. It would be a shame for Prisma not to keep this feature because other packages do not support the typescript flag. I patch a few 3rd party packages internally to support this and will be making PRs in different repos in order to help support this. For packages I could not patch, I used object spreading (i.e. `...userId ? { userId } : {}`, which isn't ideal, but the benefits of having this feature with Prisma outweigh the annoyance of the spreading 10x.
3. `exactOptionalPropertyTypes` is great as a feature of Typescript that I would be very sad not to be able to use due to Prisma dropping support for it!
4. There's a small bug when using this in combo w/ query extensions (#25845) but was able to find a workout pretty quick
5. It would be great to have `isSkip` exported from `@prisma/client/runtime/library` in future releases so developers can build client extensions that support `Prisma.skip` as a first-class citizen
6. It would be great if the skip lived on the client instance `prisma.skip` vs. `Prisma.skip` and didn't require another import to use

**Comment by ace130-github** [comment]
TLDR: 
* Is the preview feature enabled by default in Prisma 6?
* Have the semantics of  `undefined` changed from 5.19 to 5.20 if preview feature is **not** enabled?

Full text:
A little feedback concerning the [documentation](https://www.prisma.io/docs/orm/prisma-client/special-fields-and-types/null-and-undefined#strict-undefined-checks-preview-feature) around this feature:
1. As of now, the docs still say: "This new behavior is intended to become the default in Prisma ORM 6.". I think this is outdated, since it's not the default in Prisma 6.
2. Documentation around this feature is ambigous with respect to the behaviour of Prisma 5.20.0+ when the preview feature is *not* enabled: The warning at the top says: "**Prior to Prisma ORM 5.20.0**, undefined was treated as a special value that would not be included in the generated query." (emphasis added by me), followed by the recommendation to enable the preview feature with 5.20.0+. This leaves me wondering if 5.20.0+ introduces a serious change in behavoiur compared with 5.19. This is reinforced by the documentation of the "legacy behaviour" further down the page, which says that it's (only) relevant if I'm using 5.19 or earlier.

Would it be appropriate to file an issue against the documentation?

**Reply by jharrell** [reply]
Hi @ace130-github, yes, please open a docs issue and we'll resolve that :)

**Comment by MonkeeMan1** [comment]
this new preview feature is a step in the right direction and I am very happy to see it in preview. However, I do have some issues with its implementation:

I ran into a bug today because of this, typescript doesnt detect the issue, at all. Opening up alot of doors to run time errors. 
```ts
dashboard:dev: → 283     await __TURBOPACK__imported__module__$5b$project$5d2f$apps$2f$dashboard$2f$src$2f$server$2f$db$2e$ts__$5b$app$2d$route$5d$__$28$ecmascript$29$__["db"].account.update({
dashboard:dev:             where: {
dashboard:dev:               id: "123"
dashboard:dev:             },
dashboard:dev:             data: {
dashboard:dev:               access_token: "123",
dashboard:dev:               refresh_token: "123",
dashboard:dev:               user: {
dashboard:dev:                 update: {
dashboard:dev:                   name: "123",
dashboard:dev:                   image: undefined,
dashboard:dev:                          ~~~~~~~~~
dashboard:dev:                   email: "123@gmail.com"
dashboard:dev:                 }
dashboard:dev:               },
dashboard:dev:               updatedAt: new Date("2025-03-03T16:49:24.928Z")
dashboard:dev:             }
dashboard:dev:           })
dashboard:dev:
dashboard:dev: Invalid value for argument `update`: explicitly `undefined` values are not allowed.
```
User model showing `image` is completely optional, so in my opional `undefined` should be an allowed value, or at the very least the types should tell us undefined isn't allowed, so we would have to use `null` instead.
```ts
model User {
  id               String             @id @default(cuid())
  name             String?
  email            String?            @unique
  emailVerified    DateTime?
  image            String?
```

![image](https://github.com/user-attachments/assets/6f787c3d-9cc2-4813-8429-1a7ae7132973)

I think the best solution would be for the strict undefined check to only be done in where clauses, not in the actual data for the row creation.

**Comment by renatocron** [comment]
Hello, I want to ask if this will be enforced later (eg: prisma 7) or if this gonna be facultative even after preview.
As I upgraded to prisma 6 and without adding the strictUndefinedChecks the code does run in prisma 6 (I thought it would block during build time when I read the release for 5.20)
If this gonna be required later, would be nice to have access to Prisma.Skip without having to add strictUndefinedChecks so we could do a partial move, because I have a codebase with more than 1300 points just for findAll, even with claude compute this gonna take a few days to validate

**Comment by basememara** [comment]
Is this enabled by default in v6? The thread and docs give conflicting info.

Also, this _must_ be a compile-time error not runtime. In large enterprise apps, runtime errors like this are nearly impossible to catch. If this becomes the default, there needs to be an opt-out to avoid surprise bugs during future upgrades. This is painful although I fully agree with the intent of the feature.

Is there any hope this can be a compile-time error?

**Reply by basememara** [reply]
..come to think of it, this would cause surprise bugs even during development not just upgrades. Compile-time errors are critical for a feature like this.

**Comment by leighman** [comment]
Is this enabled in v6? As above the 5.22 release notes imply it would be but there's nothing in the 6 release notes.
Where actually is `Prisma.skip` ? I'm probably being stupid but I can't import it?

**Reply by renatocron** [reply]
It's not enabled on v6 unless you enabled on schema, but still not fatal on compile, just on runtime

**Comment by niallpaterson** [comment]
Anybody know how this feature interacts with where clauses in upsert operations?
